src/main/java/com/funeat/product/persistence/ProductRepositoryImpl.java

```diff
@@ -0,0 +1,86 @@
+package com.funeat.product.persistence;
+
+import com.funeat.product.domain.Product;
+import com.funeat.tag.domain.Tag;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.PersistenceContext;
+import jakarta.persistence.TypedQuery;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+@Repository
+public class ProductRepositoryImpl implements ProductRepositoryCustom {
+
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    @Override
+    public List<Product> searchProductsByTopTagsFirst(final Long tagId, final Pageable pageable) {
+        final String jpql = """
+                SELECT DISTINCT p\s
+                FROM Product p
+                WHERE p.id IN (
+                    SELECT p2.id\s
+                    FROM Product p2\s
+                    JOIN p2.reviews r2\s
+                    JOIN r2.reviewTags rt2\s
+                    WHERE rt2.tag.id = :tagId\s
+                      AND rt2.tag.id IN (
+                          SELECT rt3.tag.id\s
+                          FROM Review r3\s
+                          JOIN r3.reviewTags rt3\s
+                          WHERE r3.product.id = p2.id\s
+                          GROUP BY rt3.tag.id\s
+                          ORDER BY COUNT(rt3.tag.id) DESC\s
+                          LIMIT 3
+                      )
+                )
+                ORDER BY p.id DESC
+                """;
```

Comment:
```suggestion
        final String jpql = """
                SELECT DISTINCT p 
                FROM Product p 
                WHERE p.id IN ( 
                    SELECT p2.id 
                    FROM Product p2 
                    JOIN p2.reviews r2 
                    JOIN r2.reviewTags rt2 
                    WHERE rt2.tag.id = :tagId 
                      AND rt2.tag.id IN ( 
                          SELECT rt3.tag.id 
                          FROM Review r3 
                          JOIN r3.reviewTags rt3 
                          WHERE r3.product.id = p2.id 
                          GROUP BY rt3.tag.id 
                          ORDER BY COUNT(rt3.tag.id) DESC 
                          LIMIT 3 
                      ) 
                ) 
                ORDER BY p.id DESC
                """;
```

여기에 `\s`를 사용한 이유가 있나요?? 문제 있는게 아니라면 그냥 스페이스바로 수정하는게 더 코드 보기 편한 것 같아요